### PR TITLE
*: improve performance

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

improve performance as we did in _release-1.0_ branch.

### What is changed and how it works?

- manually pick code from _release-1.0_, including:
    - #588 
    - #590 
    - #594 
    - #597 
- fix some code conflicts.

image

image

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@WangXiangUSTC @lichunzhu PTAL

File: syncer/dml.go [hunk 1]
```diff
@@ -217,44 +214,79 @@ func genDeleteSQLs(param *genDMLParam) ([]string, [][]string, [][]interface{}, e
...
+// genDeleteSQL generates a `DELETE FROM` SQL with `WHERE`.
+func genDeleteSQL(table string, value []interface{}, columns []*model.ColumnInfo, indexColumns *model.IndexInfo) (string, []interface{}) {
+	whereColumns, whereValues := columns, value
+	if indexColumns != nil {
+		whereColumns, whereValues = getColumnData(columns, indexColumns, value)
 	}
-	return strings.Join(values, ",")
+
+	var buf strings.Builder
+	buf.Grow(1024)
```
Comment (lichunzhu) [inline syncer/dml.go:275]: Can we estimate the size? If it's too hard, I think current implemention is also okay.

Comment (csuzhangxc) [reply syncer/dml.go:275]: a little complex to estimate the size (it may make the performance worse)

**Pull request merged**